pkg/volume/util/hostutil/hostutil_linux.go

```diff
@@ -111,11 +114,28 @@ func (hu *HostUtil) GetDeviceNameFromMount(mounter mount.Interface, mountPath, p
 // the mount path reference should match the given plugin mount directory. In case no mount path reference
 // matches, returns the volume name taken from its given mountPath
 func getDeviceNameFromMount(mounter mount.Interface, mountPath, pluginMountDir string) (string, error) {
-	refs, err := mounter.GetMountRefs(mountPath)
+	var refs []string
+	// This function is used in volume reconstruction after kubelet restart.
+	// Any error returned from here can result in a volume being marked as unused, while it is still
+	// mounted, therefore possibly resulting in the volume mounted elsewhere and corrupted.
+	// Try really hard (1 minute + spam the log a lot!) to get the mount refs on InconsistentReadError.
+	err := wait.PollImmediate(10*time.Millisecond, time.Minute, func() (bool, error) {
```

Comment:
This function is also used for unmounting the volume in some places and hence not always for reconstruction. Unmount operation is usually retried and hence need not be blocking but reconstruction is one time operation and hence reading `/proc/mounts` is kind of blocking there (i.e if we can't read `/proc/mounts` on reconstruction - something is really weird on this host).

How about we  return `IsInconsistentReadError` error from here and then we provide another version of `hostutil.GetDeviceNameFromMount`  which has longer timeout (like 2-3 mins) and we use that version if `GetDeviceNameFromMount` returns `IsInconsistentReadError`. If we still can't read `/proc/mounts` then we should either taint the node or mark it is `NotReady`. 

I am proposing tainting the node or marking as `NotReady` because crashing the kubelet will simply cause it to go in a restart loop I think (assuming systemd or something similar is managing it). The other thing is - why we can't read `/proc/mounts` consistently? Is it because machine has high CPU load or machine has too many mount points which are coming and going? If it is later then - shouldn't we prevent other pods from being scheduled on the node? Regardless - I think we need to raise some high severity alert somewhere when this happens and one way of doing that - is to taint the node.   Having said that - if we can't read `/proc/mounts` consistently because of high CPU load or some different reason then it is not clear what would be mitigation in that case and what should user do to remove the taint or to get rid of the warning we are emitting.

Reply:
hostutil.GetDeviceNameFromMountWIthRetry with timeout configuration?

Reply:
also I see mount_util has a function getDeviceNameFromMount too https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/staging/src/k8s.io/mount-utils/mount.go#L190

they have different but kind of similar implementation. just wondering why we have two such functions.

Reply:
Inconsistent read from `/proc/mount` means that the mount table was modified in between two `ioutil.ReadFile()` calls. IMO it's not worth tainting the node - we try to read the file three times and expect at least two of these reads to be the same. Kubelet just had a bad luck as other processes / goroutines are mounting/unmounting other volumes.

> This function is also used for unmounting the volume in some places and hence not always for reconstruction.

It's more complicated. This function is used only for reconstruction, however, I've found GetDeviceNameFromMount copy in k8s.io/mount-utils/mount.go that's used by some drivers to unmount volumes. I'll see if they can be unified and how to add some `MustGetDeviceNameFromMount` for reconstruction.

Reply:
`hostutil.GetDeviceNameFromMount` and `mounter.GetDeviceNameFromMount` have very different semantics, the first one should be renamed to `GetRelativeGlobalVolumePath`, it returns relative dir from say `/var/lib/kubelet/plugins/kubernetes.io/aws-ebs/`. The later one returns a real device in `/dev`.

Reply:
Renamed `hostutil.GetDeviceNameFromMount` to `hostutil.GetRelativeGlobalMountPath`, which describes better what the function does.